ci/snapcraft.sh

```diff
@@ -0,0 +1,130 @@
+#!/bin/sh
+
+set -e
+
+# This script encapsulates the job of preparing snapcract configuration for the
+# rustup snap.  We construct either an edge snap (builds on PRs or master)
+# or a beta snap if built from a non-master branch.
+
+# If built on master, we should publish the edge snap
+# If built on stable, we should publish the beta snap
+
+# To parameterise the build, the following environment variables are needed:
+# DO_SNAP=1  <-- without this, we don't run
+# SNAP_EDGE=1 <-- with this, we build an edge/devel snap, otherwise a release snap
+# SNAP_ARCH=... <-- without this we cannot build, this is the snap architecture we're building
+
+if [ -z "$DO_SNAP" ]; then
+  echo "SNAP: Not attempting to snap"
+  exit 0
+fi
+
+if [ -z "$SNAP_ARCH" ]; then
+  echo "SNAP: Unable to generate snap, architecture unset"
+  exit 1
+else
+  echo "SNAP: Constructing snap for architecture $SNAP_ARCH"
+fi
+
+if [ -n "$SNAP_EDGE" ]; then
+  echo "SNAP: Generated snap will be an grade:devel snap"
+  GRADE=devel
+else
+  echo "SNAP: Generated snap will be a grade:stable snap"
+  GRADE=stable
+fi
+
+
+# This is the list of proxies which the snap will contain.  This needs to match
+# the set of aliases which the snap has been allocated.  If we add to this
+# list but don't get the aliases added, then people won't see the proxy except
+# as rustup.$PROXY.  If we have an alias which is not a supported proxy name
+# then rustup might get sad.
+
+PROXIES="cargo cargo-clippy cargo-fmt cargo-miri clippy-driver rls rustc rustdoc rustfmt"
+
+# From now on, things should be automagic
+
+VERSION=$(grep "^version" Cargo.toml | head -1 | cut -d\" -f2)
+
+rm -rf snapcraft
+
+mkdir -p snapcraft/snap
+
+cd snapcraft
+
+cat > snap/snapcraft.yaml <<SNAP
+name: rustup
+version: $VERSION
+confinement: classic
+base: core18
+architectures:
+  - build-on: [amd64]
+    run-on: [$SNAP_ARCH]
+description: |
+  Rustup is the Rust Language's installer and primary front-end.  You probably
+  want this if you want to develop anything written in Rust.
+SNAP
+
+if [ -n "$SNAP_EDGE" ]; then
+  cat >> snap/snapcraft.yaml <<SNAP
+
+  Please note, this snap is experimental and functionality cannot be guaranteed
+  to be consistent with released snaps.
+summary: "EXPERIMENTAL: The Rust Language Installer"
+grade: $GRADE
+SNAP
+else
+  cat >> snap/snapcraft.yaml <<SNAP
+summary: "The Rust Language Installer"
+grade: $GRADE
+SNAP
+fi
+
+cat >> snap/snapcraft.yaml <<SNAP
+parts:
+  acquire:
+    plugin: dump
+    source: inputs
+    prime:
+     - "-rustup-init"
+  link:
+    plugin: nil
+    build-attributes: [no-patchelf]
+    after: [acquire]
+    override-pull: |
+      cp \$SNAPCRAFT_STAGE/rustup-init .
+    override-build: |
+      mkdir -p \$SNAPCRAFT_PART_INSTALL/bin
+      cp rustup-init \$SNAPCRAFT_PART_INSTALL/bin/rustup
+SNAP
+
+for PROXY in $PROXIES; do
+  cat >> snap/snapcraft.yaml <<SNAP
+      ln \$SNAPCRAFT_PART_INSTALL/bin/rustup \$SNAPCRAFT_PART_INSTALL/bin/$PROXY
```

Comment:
Is there any particular reason for using hard links over symlinks here? I'm curious to know if they get preserved as hard links in the resultiing squashfs file.

If they don't get preserved, then symlinks could give a noticeable reduction in package size.

Reply:
The hardlinks are there because that's what `rustup` normally does and I wasn't interested in risking things.  I believe squashfs does hard links just fine, so unless the snapcraft settings fail to preserve them we ought to be fine.  If you're worried we can try switching to symlinks once the rest is sorted.

Reply:
I know squashfs can handle them.  I was more curious about whether snapcraft itself was preserving them as the data is copied from the part's install dir, to the stage dir, and finally to the prime dir.

This may not even be a problem in the end, so I agree that investigating this shouldn't be a blocker for landing this PR.

Reply:
:+1:

Reply:
I've checked and the snap I end up with on x86_64 linux is 2.7M which is definitely not a multiple of the binary size, so either squashfs has noticed they're all the same and compressed well, or else the hardlinks worked :D